### DISCUSSION TITLE
Is there a way how to extend an existing panel.

### DISCUSSION THREAD

**Discussion by BatuhanCakir** [foxglove/community#713] [Q&A]
HI,
So i need a custom panel that adds some funtionality to the existing 3D panel. Is there a way how i can just the the 3d panel as a base and just add stuff instead of doing it from scratch.

Thank you in advance

**Comment by defunctzombie** [comment]
The only way to have to extend an existing panel is via message converter extensions: https://foxglove.dev/docs/studio/extensions/extension-context#registermessageconverter These allow you to convert a custom message format into a well-known format for visualization in a panel.

There is no other mechanism for extending existing panels. What are you looking to change/add to the 3d panel? Sharing your use case could help us understand what kind of extension points to think about.

**Reply by BatuhanCakir** [reply]
I want to add the feature that you add waypoints in the panel. 
Like you can click in the 3d panel on a random point and it gets visualized.
And with that some UI Buttons for deleting all the Points and saving the points(sending it to a ros roboter)

**Reply by defunctzombie** [reply]
The 3d panel has a publish feature. If your robotics stack listens for the published messages it can then respond by publishing markers or annotations for the location that it received and that will be visualized on the panel. The nice property of this design is that you see what the robot actually received and acknowledged.

**Comment by BatuhanCakir** [comment]
Ok thank you for your help.

Am 16.05.2023 um 17:53 schrieb Roman Shtylman ***@***.***>:

﻿

The 3d panel has a publish feature. If your robotics stack listens for the published messages it can then respond by publishing markers or annotations for the location that it received and that will be visualized on the panel. The nice property of this design is that you see what the robot actually received and acknowledged.

**Comment by bramanat94** [comment]
Hai can i add layers to the custom panel, if possible how? @defunctzombie . I need to add many layers to the 3d panel, so that i can display all the information in a single panel with different layers.
